lib/mimemagic.rb

```diff
@@ -122,11 +122,16 @@ def self.magic_match_io(io, matches, buffer)
     matches.any? do |offset, value, children|
       match =
         if Range === offset
-          io.read(offset.begin, buffer)
+          io.seek(offset.begin)
           x = io.read(offset.end - offset.begin + value.bytesize, buffer)
           x && x.include?(value)
+        elsif offset.negative?
+          max_offset = [0 - io.size, offset].max
```

Comment:
This calculation is fundamental to not overflow the file pointer. For example, if the file has 1K, and I seek 1.1K from end to begin, it will go out of file.